core/arch/riscv/include/kernel/thread_private_arch.h

```diff
@@ -0,0 +1,110 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022-2023 NXP
+ */
+
+#ifndef __KERNEL_THREAD_PRIVATE_ARCH_H
+#define __KERNEL_THREAD_PRIVATE_ARCH_H
+
+#ifndef __ASSEMBLER__
+
+#include <kernel/thread.h>
+
+#define STACK_TMP_OFFS		0
+
+#if defined(__clang__) && !defined(__OPTIMIZE_SIZE__)
+#define STACK_TMP_SIZE		(4096 + STACK_TMP_OFFS + CFG_STACK_TMP_EXTRA)
+#else
+#define STACK_TMP_SIZE		(2048 + STACK_TMP_OFFS + CFG_STACK_TMP_EXTRA)
+#endif
+
+#define STACK_THREAD_SIZE	(8192 + CFG_STACK_THREAD_EXTRA)
+
+#if TRACE_LEVEL > 0
+#define STACK_ABT_SIZE		3072
+#else
+#define STACK_ABT_SIZE		1024
+#endif
+
+#define STACK_CHECK_EXTRA	0
+
+#define THREAD_RPC_NUM_ARGS     4
+
+#define	TRAP_MODE_KERNEL	0
+#define	TRAP_MODE_USER		1
+
+struct thread_user_mode_rec {
+	unsigned long ctx_regs_ptr;
+	unsigned long exit_status0_ptr;
+	unsigned long exit_status1_ptr;
+	unsigned long pad;
+	/*
+	 * x[] is used to save registers for user/kernel context-switching
+	 * 0-3: ra-tp
+	 * 4-6: s0-s1
+	 * 6-15: s2-s11
+	 */
+	unsigned long x[16];
+};
+
+#endif /*__ASSEMBLER__*/
```

Comment:
A summary of the issue:
```
#include <stdint.h>

#ifdef __ASSEMBLER__
#define SHIFT_U64(v, shift)	((v) << (shift))
#else
#define SHIFT_U64(v, shift)	((uint64_t)(v) << (shift))
#endif

#define TO_STR(x) _TO_STR(x)
#define _TO_STR(x) #x

#define CSR_MODE_OFFSET	1

#define CSR_MODE_BITS	SHIFT_U64(CSR_MODE_OFFSET, 8)

#define CSR_XSTATUS		(CSR_MODE_BITS | 0x000)
#define CSR_XIE			(CSR_MODE_BITS | 0x004)

#define clear_csr(reg, bit) ({ unsigned long __tmp; \
  asm volatile ("csrrc %0, " TO_STR(reg) ", %1" : "=r"(__tmp) : "rK"(bit)); \
  __tmp; })
 
int main()
{
	clear_csr(CSR_XSTATUS, CSR_XIE);
	return 0;
}

/*
maro-vm:~/test$ riscv64-linux-gnu-gcc test.c -o test
test.c: Assembler messages:
test.c:25: Error: found '(', expected: ')'
test.c:25: Error: found '(', expected: ')'
test.c:25: Error: Instruction csrrc requires absolute expression
test.c:25: Error: found '(', expected: ')'
test.c:25: Error: found '(', expected: ')'
test.c:25: Error: Instruction csrrc requires absolute expression
test.c:25: Error: illegal operands `csrrc a5,(((uint64_t)(1)<<(8))|0x000),a5'
maro-vm:~/test$ riscv64-linux-gnu-gcc -S -o- test.c
        .file   "test.c"
        .option nopic
        .text
        .align  1
        .globl  main
        .type   main, @function
main:
        add     sp,sp,-32
        sd      s0,24(sp)
        add     s0,sp,32
        li      a5,260
#APP
# 25 "test.c" 1
        csrrc a5, (((uint64_t)(1) << (8)) | 0x000), a5
# 0 "" 2
#NO_APP
        sd      a5,-24(s0)
        li      a5,0
        mv      a0,a5
        ld      s0,24(sp)
        add     sp,sp,32
        jr      ra
        .size   main, .-main
        .ident  "GCC: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0"
maro-vm:~/test$
*/
```

Reply:
The problem is that `clear_csr()` is a macro instead of a static inline function. It wouldn't be too hard to fix the macro, but static inline functions are usually more robust.

Reply:
I'll move CSR handling functions to riscv.h un define them as inline functions.

Reply:
OR better declare new ones since this header is auoto-generated clear_csr() -> csr_clear() ?

Reply:
How will one know which one to use? Anyway, It's not a big deal. If you prefer it this way, it's okay.

Reply:
@jenswi-linaro, the problem is that the encoding.h where these CSR functions is automatically generated, tracking changes will be hard, what bout fixing the marco ?

Reply:
Can't this be fixed upstream instead?

Reply:
@jforissier these macros are used as they are in all other projects.

Reply:
Yes and what I'm saying is that by replacing them with static inline functions, nothing should break and everyone would benefit from being able to use C syntax arguments. Do you have any particular reason to think such an update would be rejected upstream?

Reply:
I'll give it a try.